receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/metadata/generated_metrics.go

```diff
@@ -17,81 +17,180 @@
 	"go.opentelemetry.io/collector/model/pdata"
 )
 
 // Type is the component type name.
 const Type config.Type = "cpu"
 
-// MetricIntf is an interface to generically interact with generated metric.
-type MetricIntf interface {
-	Name() string
-	New() pdata.Metric
-	Init(metric pdata.Metric)
+type metricMetadata struct {
+	name        string
+	enabled     bool
+	description string
+	unit        string
+	dataType    pdata.MetricDataType
+	isMonotonic bool
+	temporality pdata.MetricAggregationTemporality
 }
 
-// Intentionally not exposing this so that it is opaque and can change freely.
-type metricImpl struct {
-	name     string
-	initFunc func(pdata.Metric)
+// metricBuilder provides an interface for scrapers to report metrics while taking care of all the transformations
+// required to produce metric representation defined in metadata and user configuration.
+type metricBuilder struct {
+	metadata        metricMetadata
+	config          MetricConfig
+	metric          pdata.Metric
+	initialCapacity int
+	startTime       pdata.Timestamp
 }
 
-// Name returns the metric name.
-func (m *metricImpl) Name() string {
-	return m.name
+// MetricBuilderOption applies changes to default metrics builder.
+type MetricBuilderOption func(metricBuilder)
+
+// WithStartTime sets startTime on the metrics builder.
+func WithStartTime(startTime pdata.Timestamp) MetricBuilderOption {
+	return func(mb metricBuilder) {
+		mb.startTime = startTime
+	}
 }
 
-// New creates a metric object preinitialized.
-func (m *metricImpl) New() pdata.Metric {
-	metric := pdata.NewMetric()
-	m.Init(metric)
-	return metric
+// WithInitialCapacity sets initial capacity for metric data points.
+func WithInitialCapacity(initialCapacity int) MetricBuilderOption {
+	return func(mb metricBuilder) {
+		mb.initialCapacity = initialCapacity
+	}
 }
 
-// Init initializes the provided metric object.
-func (m *metricImpl) Init(metric pdata.Metric) {
-	m.initFunc(metric)
+func newMetricBuilder(metadata metricMetadata, config MetricConfig, options ...MetricBuilderOption) metricBuilder {
+	mb := metricBuilder{
+		metadata: metadata,
+		config:   config,
+	}
+	if !mb.Enabled() {
+		return mb
+	}
+
+	mb.createMetric()
+	mb.startTime = pdata.NewTimestampFromTime(time.Now())
+
+	for _, op := range options {
+		op(mb)
+	}
+	return mb
 }
 
-type metricStruct struct {
-	SystemCPUTime MetricIntf
+// Enabled identifies whether the metrics should be collected or not.
+func (mb *metricBuilder) Enabled() bool {
+	if mb.config.Enabled != nil {
+		return *mb.config.Enabled
+	}
+	return mb.metadata.enabled
+}
+
+// EnsureDataPointsCapacity ensures metric data points slice capacity.
+func (mb *metricBuilder) EnsureDataPointsCapacity(cap int) {
+	if !mb.Enabled() {
+		return
+	}
+	mb.metric.Sum().DataPoints().EnsureCapacity(cap)
+}
+
+// Reset resets the metric builder startTime and removes previous/current metric state.
+func (mb metricBuilder) Reset(options ...MetricBuilderOption) {
```

Comment:
Probably not needed in the first version, but good to know we can support this.

Reply:
Ok, sounds good